nixos/self-deploy: init

lambdadog opened the pull request:
Add `self-deploy` service to facilitate continuous deployment of a NixOS system derivation from a git repository to the local system. This is based on an internal NixOS module of the same name we've been using at Awake Security that we've generalized for open source use.

###### Motivation for this change

At Awake Security we've found our internal version of this service useful in ensuring infrastructure is always up to date and felt it could be useful for others.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
nixos/self-deploy: init

Add `self-deploy` service to facilitate continuous deployment of NixOS
configuration from a git repository.

evanrelf commented on nixos/modules/services/system/self-deploy.nix:
```diff
@@ -0,0 +1,174 @@
+      '';
+    };
+
+    branch = lib.mkOption {
+      type = lib.types.str;
+
+      default = "master";
+
+      description = ''
+      Branch to track
+
+      You can also specify a revision or tag, too, if you want to pin this
+      machine to a specific commit.
+      '';
+    };
```
This `branch` value is used in a `git fetch` command, but `git fetch` requires a ref, not a rev. So I wouldn't say you can specify a revision.

evanrelf replied:
@lambdadog I'm familiar with this issue because I ran into it when writing `manual-deploy`.

If you tweak the options to more closely match what I have with `manual-deploy`'s CLI options, where the user explicitly says whether they're specifying a rev or a ref, then you can handle both cases.

You can't `fetch` a specific `rev`, so you'd just fetch everything from the remote, and then `checkout` that `rev`.

lambdadog replied:
I honestly don't feel it's necessary to deeply support the use-case of fetching a specific rev, seeing as the use-case for `self-deploy` is continuous deployment and a single revision isn't going to change in a way that requires regular redeployment. I'll change the wording of the description, but unless you can think of a compelling usecase for it I don't see myself implementing an alternate path just for this.

evanrelf replied:
Sounds good 👍  I agree a specific rev isn't the point of this service.

evanrelf commented on nixos/modules/services/system/self-deploy.nix:
```diff
@@ -0,0 +1,174 @@
+      default = "hourly";
+
+      description = ''
+      The schedule on which to run the `self-deploy` service. Format
+      specified by `systemd.time 7`.
+
+      This value can also be a list of `systemd.time 7` formatted
+      strings, in which case the service will be started on multiple
+      schedules.
```
I think `systemd.time(7)` with the parenthesis is more correct, or you could say `man 7 systemd.time` if you want to be more helpful.

lambdadog replied:
I was just matching the wording used in [`systemd.services.<name>.startAt`](https://search.nixos.org/options?channel=20.09&show=systemd.services.%3Cname%3E.startAt&from=0&size=50&sort=relevance&query=systemd.services.%3Cname%3E.startAt) since this value is just copied directly into it.

evanrelf replied:
Ah okay, then disregard my comment 👍

evanrelf commented on nixos/modules/services/system/self-deploy.nix:
```diff
@@ -0,0 +1,174 @@
+      environment.GIT_SSH_COMMAND = lib.mkIf (!(isNull cfg.sshKeyFile))
+        "${pkgs.openssh}/bin/ssh -i ${lib.escapeShellArg cfg.sshKeyFile}";
+
+      serviceConfig.X-RestartIfChanged = false;
```
I'd use `systemd.services.<name>.restartIfChanged` instead.

Gabriel439 commented on nixos/modules/services/system/self-deploy.nix:
```diff
@@ -0,0 +1,174 @@
+        }} ${lib.escapeShellArg cfg.repository} ${repositoryDirectory}
+      fi
+
+      # Ensure that if cfg.repository is changed, origin is updated
+      ${gitWithRepo} remote set-url origin ${lib.escapeShellArg cfg.repository}
+
+      ${gitWithRepo} fetch origin ${lib.escapeShellArg cfg.branch}
```
I think you can simplify this and make it more "pure" by doing this:

```suggestion
      ${gitWithRepo} fetch ${lib.escapeShellArg cfg.repository} ${lib.escapeShellArg cfg.branch}
```

lambdadog replied:
Hmm, so it doesn't need to be added as a remote to fetch from it? interesting

I wonder if it would be better to create a bare git repository instead of cloning in that case

lambdadog replied:
Confirmed the fresh git repo works, going to switch to that

Commit:
nixos/self-deploy: fixups

Changes:
 - Correct `services.self-deploy.branch` description
 - Use `systemd.services.<name>.restartIfChanged`
 - Switch to using empty git repo and fetching without adding remote
 - Move `systemctl reboot` to after cleanup tasks

SuperSandro2000 commented:
@ofborg eval

SuperSandro2000 commented on nixos/modules/services/system/self-deploy.nix:
```diff
@@ -0,0 +1,167 @@
+let
+  cfg = config.services.self-deploy;
+
+  workingDirectory = "/var/lib/self-deploy";
```
Maybe ``/var/lib/nixos-self-deploy`` would be a better directory. Not sure though.

lambdadog replied:
Hmm. I doubt it will be necessary, but it might be a good idea to do anyway. Namespacing it with a `nixos-` prefix will all but ensure there's no chance of a collision.

SuperSandro2000 commented on nixos/modules/services/system/self-deploy.nix:
```diff
@@ -0,0 +1,167 @@
+    systemd.services.self-deploy = {
+      wantedBy = [ "multi-user.target" ];
+
+      requires = lib.mkIf (!(isPathType cfg.repository)) [ "network-online.target" ];
```
If the script runs by a cron we don't really need network at startup, right?

lambdadog replied:
It's probably best to make it explicit that network is needed regardless. Probably will never be necessary, but there's something to be said for correctness.

SuperSandro2000 replied:
Also I am not sure if this requires network target additionally.

SuperSandro2000 reviewed (commented):
Can you format the file with nixpkgs-fmt? I don't want to write a suggestion for every little indentation.

Also some of these options could require an example value like sshKeyFile, nixArgs or startAt.

Commit:
nixos/self-deploy: make recommended changes

Changes:
 - /var/lib/{self-deploy->nixos-self-deploy}
 - run nixpkgs-fmt

lambdadog commented:
@SuperSandro2000 made your changes, other than the network-online requirement, which I'd prefer to keep for correctness

lambdadog commented:
@SuperSandro2000 made proposed changes other than `network-online.target` requirement, since I would argue it's "more correct" to require it, even if it strictly shouldn't be necessary.

SuperSandro2000 commented:
> @SuperSandro2000 made proposed changes other than `network-online.target` requirement, since I would argue it's "more correct" to require it even if it strictly shouldn't be necessary.

That has nothing to do with being correct or not.  Network-online delays the start unecessary and it also does not mean you have internet connection necessarily.

Gabriel439 commented:
@SuperSandro2000: The `self-deploy` service runs hourly by default, so if the first run fails (due to the network not being online), then the first run that succeeds might be as late as an hour after boot.  Rather than optimizing for "time to first (possibly failed) run" we should be optimizing for "time to first successful run" and adding the `network-online.target` helps for the latter criterion.

Yes, `network-online.target` does not completely guarantee that the network being available, but it's still a decent approximation and improves the likelihood of success.

lambdadog commented:
As Gabriel mentions, my understanding is that, while the `network-online` target doesn't ensure that you necessarily have an internet connection, the inverse (excluding exceptions like your intentional disabling of it) *is* true. When `network-online` is down you can fundamentally assume that you have no internet connection.

By this argument, any time that `self-deploy` runs when `network-online` is down on a system where `network-online` isn't specifically disabled is ensured to fail. It seems like by removing it you're specifically optimizing for a case that will 100% produce a failed run (unless the network comes up between the service starting and the fetch, and this is quite unlikely IMO, since it's near the first thing `self-deploy` does).

Gabriel439 commented:
@SuperSandro2000: Do you have any other feedback before we merge this?

SuperSandro2000 commented:
not right now

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-use-services-self-deploy/20696/1
